### ISSUE TITLE
chef-client local mode will throw a 404 every time a template action is attempted

### ISSUE THREAD

**Issue by Ryuzavi** [chef/chef#6283]
# Description

After updating to the latest chefdk we've noticed that a local chef-client run will fail every time it attempts to copy in a template file with 404 http error.

One weird quirk I've seen is when I run a specific cookbook on it's own that does a template it's fine.  This seems to only happen when I run my full run across multiple cookbooks.  It's always the first template action attempt that fails.

This works for us on the previous chef-client version 12.18.31.

Debug output suggests it's a call to chef zero that's the problem.  Specifically the `PATH_INFO` string appears to be empty for the offedning call (other calls seem to work fine)

# ChefDK Version

2.0.28

# Platform Version

Centos 7.3

## Replication Case

Having a couple of cookbooks where one is dependent and calls the others.  Have one of the dependents attempt to create a template,  Run the cookbook with chef-client local mode.

## Client Output

```
  * template[/var/www/html/index.php] action create

    ================================================================================
    Error executing action `create` on resource 'template[/var/www/html/index.php]'
    ================================================================================

    Net::HTTPServerException
    ------------------------
    404 "Not Found"

    Resource Declaration:
    ---------------------


    Compiled Resource:
    ------------------
    # Declared in /root/.chef/local-mode-cache/cache/cookbooks/splash-page/recipes/install-splash-page.rb:34:in `from_file'

    template("/var/www/html/index.php") do
      action [:create]
      default_guard_interpreter :default
      source "index.php.erb"
      declared_type :template
      cookbook_name "splash-page"
      recipe_name "install-splash-page"
      owner "user"
      group "apache"
      mode "0755"
      path "/var/www/html/index.php"
      verifications []
    end

    System Info:
    ------------
    chef_version=13.2.20
    platform=centos
    platform_version=7.3.1611
    ruby=ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
    program_name=chef-client worker: ppid=1588;start=17:08:09;
    executable=/opt/chefdk/bin/chef-client
```

Specific error from debug output:
```
[2017-07-19T15:35:35+01:00] DEBUG: #<ChefZero::RestRequest:0x00000006d710f8
 @env=
  {"SCRIPT_NAME"=>"",
   "SERVER_NAME"=>"localhost",
   "REQUEST_METHOD"=>"GET",
   "PATH_INFO"=>"",
   "QUERY_STRING"=>nil,
   "SERVER_PORT"=>1,
   "HTTP_HOST"=>"localhost:1",
   "HTTP_X_OPS_SERVER_API_VERSION"=>"2",
   "rack.url_scheme"=>"chefzero",
   "rack.input"=>#<StringIO:0x00000006d71148>},
 @rest_base_prefix=["organizations", "chef"],
 @rest_path=["organizations", "chef"]>

[2017-07-19T15:35:35+01:00] DEBUG: #<ChefZero::RestErrorResponse: 404: Object not found: chefzero://localhost:1/org>
```

## Stacktrace

```
Generated at 2017-07-19 17:08:39 +0100
Net::HTTPServerException: template[/var/www/html/index.php] (splash-page::install-splash-page line 34) had an error: Net::HTTPServerException: 404 "Not Found"
/opt/chefdk/embedded/lib/ruby/2.4.0/net/http/response.rb:122:in `error!'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/http.rb:239:in `streaming_request'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/cookbook/remote_file_vendor.rb:66:in `get_filename'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/cookbook_version.rb:281:in `preferred_filename_on_disk_location'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template_finder.rb:40:in `find'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template/content.rb:32:in `template_location'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template.rb:43:in `block in define_resource_requirements'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/mixin/why_run.rb:314:in `assert'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template.rb:41:in `define_resource_requirements'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider.rb:158:in `run_action'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource.rb:591:in `run_action'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:70:in `run_action'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:98:in `block (2 levels) in converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:98:in `each'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:98:in `block in converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:97:in `converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:715:in `block in converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:710:in `catch'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:710:in `converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:749:in `converge_and_save'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:286:in `run'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:291:in `block in fork_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:279:in `fork'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:279:in `fork_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:244:in `block in run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:232:in `run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:469:in `sleep_then_run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:458:in `block in interval_run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:457:in `loop'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:457:in `interval_run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:441:in `run_application'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:59:in `run'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:263:in `load'
/usr/bin/chef-client:263:in `<main>'

>>>> Caused by Net::HTTPServerException: 404 "Not Found"
/opt/chefdk/embedded/lib/ruby/2.4.0/net/http/response.rb:122:in `error!'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/http.rb:239:in `streaming_request'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/cookbook/remote_file_vendor.rb:66:in `get_filename'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/cookbook_version.rb:281:in `preferred_filename_on_disk_location'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template_finder.rb:40:in `find'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template/content.rb:32:in `template_location'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template.rb:43:in `block in define_resource_requirements'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/mixin/why_run.rb:314:in `assert'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider/template.rb:41:in `define_resource_requirements'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/provider.rb:158:in `run_action'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource.rb:591:in `run_action'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:70:in `run_action'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:98:in `block (2 levels) in converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:98:in `each'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:98:in `block in converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/runner.rb:97:in `converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:715:in `block in converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:710:in `catch'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:710:in `converge'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:749:in `converge_and_save'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/client.rb:286:in `run'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:291:in `block in fork_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:279:in `fork'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:279:in `fork_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:244:in `block in run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:232:in `run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:469:in `sleep_then_run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:458:in `block in interval_run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:457:in `loop'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:457:in `interval_run_chef_client'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application/client.rb:441:in `run_application'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/lib/chef/application.rb:59:in `run'
/opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/chef-13.2.20/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:263:in `load'
```

**Comment by Ryuzavi** [comment]
Downgrading to chefdk 1.5.0 with chef-client 12.21.2 worked for me also.  

Thought it might be related to the chef zero socketless mode, but it still failed with the --listen flag...  
```
[2017-07-20T10:49:15+01:00] DEBUG: GET /organizations/chef
[2017-07-20T10:49:15+01:00] DEBUG: #<ChefZero::RestRequest:0x00000007653e00
 @env=
  {"SCRIPT_NAME"=>"",
   "SERVER_NAME"=>"localhost",
   "REQUEST_METHOD"=>"GET",
   "PATH_INFO"=>"",
   "QUERY_STRING"=>nil,
   "SERVER_PORT"=>8889,
   "HTTP_HOST"=>"localhost:8889",
   "HTTP_X_OPS_SERVER_API_VERSION"=>"2",
   "rack.url_scheme"=>"chefzero",
   "rack.input"=>#<StringIO:0x00000007653e50>},
 @rest_base_prefix=["organizations", "chef"],
 @rest_path=["organizations", "chef"]>

[2017-07-20T10:49:15+01:00] DEBUG: #<ChefZero::RestErrorResponse: 404: Object not found: chefzero://localhost:8889/org>
```

Maybe this is centos 7 specific or something?

**Comment by thommay** [comment]
I can't reproduce this, unfortunately. 
```
       Transferring files to <default-centos-7>
       Starting Chef Client, version 13.6.4
       Creating a new client identity for default-centos-7 using the validator key.
       resolving cookbooks for run list: ["local_2::default"]
       Synchronizing Cookbooks:
         - local_2 (0.1.0)
         - local_1 (0.1.0)
         - local_3 (0.1.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 4 resources
       Recipe: local_1::default
         * template[/tmp/foo] action create
           - create new file /tmp/foo
           - update content in file /tmp/foo from none to e3b0c4
           (no diff)
           - restore selinux security context
         * template[/tmp/foo2] action create
           - create new file /tmp/foo2
           - update content in file /tmp/foo2 from none to e3b0c4
           (no diff)
           - restore selinux security context
       Recipe: local_3::default
         * template[/tmp/foo] action create (up to date)
         * template[/tmp/foo2] action create (up to date)

       Running handlers:
       Running handlers complete
       Chef Client finished, 2/4 resources updated in 02 seconds
       Finished converging <default-centos-7> (0m19.67s).
```

Can you send a tarball of your cookbooks to me? `berks package` or `chef export`, and then send them to `thom [at] chef.io` .

**Comment by thommay** [comment]
It works in kitchen but not elsewhere? How are you distributing the cookbooks to nodes?

**Comment by Ryuzavi** [comment]
We do a git checkout of the cookbooks

**Comment by thommay** [comment]
Can you sanitize your chef config and attach that, the exact command you're running, and the output of `find <path_to_cookbooks> -name '*.rb'  -o -name '*.erb'` or similar.

**Comment by Ryuzavi** [comment]
Hi

Thanks for your help.  I was able to diagnose the issue when attempting to get you a sanitized version.

Turns out there was a custom resource committed that was specifically clearing the chef local cache and this of course meant things like templates and the like we're getting deleted.  I only found this when going through every recipe on the offending path.  I guess the cache behavior was changed in Chef 13 and upwards that this custom resource not conflicts with.  Removing the offending resource sorted the issue.

I'm not sure at this stage why this code was added in the past, maybe someone was seeing some cache not updating issues but I can investigate this further myself.  

Thanks again for you help.

Cheers,
R.
